### PR TITLE
[server] ProjectService - perm-check and testing

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

The ProjectsService does now use the new authorizer to maintain relationships and check for permissions. 
It does it in addition to any checks that happen in gitpod-server-impl.ts.
To allow checking permissions we need to pass in the userId, so the majority of file changes are that.
This PR also adds a test for the project service.

Please review https://github.com/gitpod-io/gitpod/pull/18099 first

## Related Issue(s)

Fixes WEB-521
Fixes WEB-522
Fixes WEB-524
Fixes WEB-525

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
        <li><b>🏷️ Name</b> - se-projectservice</li>
        <li><b>🔗 URL</b> - <a href="https://se-projectservice.preview.gitpod-dev.com/workspaces" target="_blank">se-projectservice.preview.gitpod-dev.com/workspaces</a>.</li>
        <li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
        <li><b>📦 Version</b> - </li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>

/hold

### CODE REVIEW

**Comment by easyCZ** [comment]
Rebase please. The parent PR landed and it now shows a difficult diff

File: components/server/src/authorization/authorizer.ts [hunk 1]
```diff
@@ -38,6 +38,22 @@ export class Authorizer {
...
+    async hasPermissionOnProject(userId: string, permission: ProjectPermission, project: Project): Promise<boolean> {
+        const req = v1.CheckPermissionRequest.create({
+            subject: subject("user", userId),
+            permission,
+            resource: objectRef("project", project.id),
+            consistency: v1.Consistency.create({
+                requirement: {
+                    oneofKind: "fullyConsistent",
+                    fullyConsistent: true,
+                },
```
Comment (easyCZ) [inline components/server/src/authorization/authorizer.ts:50]: Extract this to a constant and re-use between `hasPermissionOnProject` and `hasPermissionOnOrg`?

File: components/server/src/projects/projects-service.spec.db.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                bind(HostContextProvider).to(HostContextProviderImpl).inSingletonScope();
+
+                // auth
+                bind(SpiceDBClient)
+                    .toDynamicValue(() => {
+                        const token = uuidv4();
+                        return v1.NewClient(token, "localhost:50051", v1.ClientSecurity.INSECURE_PLAINTEXT_CREDENTIALS)
+                            .promises;
+                    })
+                    .inSingletonScope();
```
Comment (easyCZ) [inline components/server/src/projects/projects-service.spec.db.ts:65]: Can we/does it make sense, to extract this constructor and re-use between the org/project services?

File: components/server/src/projects/projects-service.spec.db.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        const repo = (await typeorm.getConnection()).getRepository(DBUser);
+        await repo.delete(owner.id);
+        await repo.delete(stranger.id);
+    });
+
+    it("should let org owners create projects", async () => {
+        const ps = container.get(ProjectsService);
+        const project = await createTestProject(ps, org, owner);
+        expect(project).to.not.be.undefined;
+    });
```
Comment (easyCZ) [inline components/server/src/projects/projects-service.spec.db.ts:95]: I think here, we can combine the unit test for a perm check, and the test for the `createProject` and check that we get the right values back.

I think in general, for the happy cases we don't have to include the "is allowed" in the test description, but for the negative cases it would make sense

File: components/server/src/projects/projects-service.ts [hunk 4]
```diff
@@ -38,19 +39,60 @@ export class ProjectsService {
...
     }
 
-    async getTeamProjects(teamId: string): Promise<Project[]> {
-        return this.projectDB.findProjects(teamId);
+    /**
+     *
+     * @param userId
+     * @param orgId
+     * @returns any user visible projects on the given org
+     */
```
Comment (easyCZ) [inline components/server/src/projects/projects-service.ts:55]: Do we really need these docstrings?

File: components/server/src/projects/projects-service.ts [hunk 5]
```diff
@@ -79,6 +121,10 @@ export class ProjectsService {
     }
 
     async getBranchDetails(user: User, project: Project, branchName?: string): Promise<Project.BranchDetails[]> {
+        if (!(await this.auth.hasPermissionOnProject(user.id, "read_info", project))) {
```
Comment (easyCZ) [inline components/server/src/projects/projects-service.ts:124]: Personally, I find the `if ( await` really hard to read.

Here, it would help me visually grep this code if it looked like this:
```suggestion
		const canRead = await this.auth.hasPermissionOnProject(user.id, "read_info", project)
        if (!canRead) {
```

File: components/server/src/projects/projects-service.ts [hunk 6]
```diff
@@ -119,6 +165,17 @@ export class ProjectsService {
         { name, slug, cloneUrl, teamId, appInstallationId }: CreateProjectParams,
         installer: User,
     ): Promise<Project> {
+        if (!(await this.auth.hasPermissionOnOrganization(installer.id, "create_project", teamId))) {
+            if (!this.auth.hasPermissionOnOrganization(installer.id, "read_info", teamId)) {
```
Comment (easyCZ) [inline components/server/src/projects/projects-service.ts:169]: missing await?

File: install/installer/pkg/components/spicedb/data/schema.yaml [hunk 7]
```diff
@@ -50,19 +50,26 @@ schema: |-
     // A subject is an editor, if:
     //  * the user is directly assigned as an editor
     relation editor: user
```
Comment (easyCZ) [inline install/installer/pkg/components/spicedb/data/schema.yaml:52]: I'm surprised we already have this extension in the model. I thought this was gonna be part of the `Fix projects` workstream

File: components/server/src/projects/projects-service.ts (hunk 6 above)
Comment (easyCZ) [reply components/server/src/projects/projects-service.ts:169]: Also goes to https://github.com/gitpod-io/gitpod/pull/18111/files#r1250981855

File: components/server/src/projects/projects-service.spec.db.ts (hunk 2 above)
Comment (svenefftinge) [reply components/server/src/projects/projects-service.spec.db.ts:65]: did that in the other PR already

File: components/server/src/projects/projects-service.ts (hunk 5 above)
Comment (svenefftinge) [reply components/server/src/projects/projects-service.ts:124]: Sure. Makes it more readable in this case, because the expression is a tad too long. I disagree though that we should generally not `await` within conditions or other expression context.

File: components/server/src/projects/projects-service.ts (hunk 6 above)
Comment (svenefftinge) [reply components/server/src/projects/projects-service.ts:169]: Good catch! FWIW I could forget the `await` when declaring the local variable just like that. Would be great to have a listing rule that errors when using Promise in the context of a condition.

File: install/installer/pkg/components/spicedb/data/schema.yaml (hunk 7 above)
Comment (svenefftinge) [reply install/installer/pkg/components/spicedb/data/schema.yaml:52]: Yes, I was also surprised. Don't know when it came in. It is not used so far.

File: components/server/src/projects/projects-service.ts (hunk 6 above)
Comment (easyCZ) [reply components/server/src/projects/projects-service.ts:169]: There's https://eslint.org/docs/latest/rules/require-await which could work. Tho I do remember we had this already. Will have a look later today.

**Review by easyCZ — APPROVED** [review]
LGTM, module comments
